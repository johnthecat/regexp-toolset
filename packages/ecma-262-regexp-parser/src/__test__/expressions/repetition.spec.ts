import { expect, describe, it } from 'vitest';
import { parseRegexp } from '../../index.js';

describe('Repetition', () => {
  describe('Range', () => {
    it('should correct work with lazy quantifier', () => {
      expect(parseRegexp(/a{2}?/)).toMatchSnapshot();
    });

    describe('Exact', () => {
      it('should parse repetition count', () => {
        expect(parseRegexp(/a{2}/)).toMatchSnapshot();
      });
      it('should parse unclosed range as chars', () => {
        expect(parseRegexp(/a{2/)).toMatchSnapshot();
      });
      it('should parse unclosed range as chars', () => {
        expect(parseRegexp(/a2}/)).toMatchSnapshot();
      });
    });

    describe('From exact to unlimited', () => {
      it('should parse repetition count', () => {
        expect(parseRegexp(/a{2,}/)).toMatchSnapshot();
      });
      it('should parse unclosed range as chars', () => {
        expect(parseRegexp(/a{2,/)).toMatchSnapshot();
      });
      it('should parse unclosed range as chars', () => {
        expect(parseRegexp(/a2,}/)).toMatchSnapshot();
      });
    });

    describe('From exact to exact', () => {
      it('should parse repetition count', () => {
        expect(parseRegexp(/a{2,4}/)).toMatchSnapshot();
      });
      it('should parse unclosed range as chars', () => {
        expect(parseRegexp(/a{2,4/)).toMatchSnapshot();
      });
      it('should parse unclosed range as chars', () => {
        expect(parseRegexp(/a2,4}/)).toMatchSnapshot();
      });

      it('should throw if range start is bigger than end', () => {
        expect(() => parseRegexp('/a{4,2}/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /a{4,2}/
             ═════
         The quantifier range is out of order"
      `);
      });
    });

    describe('Errors', () => {
      it('should detect to quantifiable tokens', () => {
        expect(() => parseRegexp('/${1}/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /\${1}/
             ═
         The preceding token is not quantifiable"
      `);
      });

      it('should throw if there is no expression before', () => {
        expect(() => parseRegexp('/{1}/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /{1}/
            ═
         There is nothing to quantify"
      `);
      });
    });
  });

  describe('None or single (?)', () => {
    it('should be parsed', () => {
      expect(parseRegexp(/a?/)).toMatchSnapshot();
    });

    it('should correct work with lazy quantifier', () => {
      expect(parseRegexp(/a??/)).toMatchSnapshot();
    });

    it('should detect to quantifiable tokens', () => {
      expect(() => parseRegexp('/$?/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /\$?/
             ═
         The preceding token is not quantifiable"
      `);
    });

    it('should throw if there is no expression before', () => {
      expect(() => parseRegexp('/?/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /?/
            ═
         There is nothing to quantify"
      `);
    });
  });

  describe('None or many (*)', () => {
    it('should be parsed', () => {
      expect(parseRegexp(/a*/)).toMatchSnapshot();
    });

    it('should correct work with lazy quantifier', () => {
      expect(parseRegexp(/a*?/)).toMatchSnapshot();
    });

    it('should detect to quantifiable tokens', () => {
      expect(() => parseRegexp('/$*/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /\$*/
             ═
         The preceding token is not quantifiable"
      `);
    });

    it('should throw if there is no expression before', () => {
      expect(() => parseRegexp('/*/')).toThrowErrorMatchingInlineSnapshot(`
        "
         ❱ /*/
            ═
         There is nothing to quantify"
      `);
    });
  });
});
