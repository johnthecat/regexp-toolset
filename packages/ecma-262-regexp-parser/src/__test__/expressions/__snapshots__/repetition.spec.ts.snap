// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Repetition > None or many (*) > should be parsed 1`] = `
{
  "body": {
    "end": 2,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": 6,
      "start": 1,
      "type": "simple",
      "value": "a",
    },
    "kind": 27,
    "quantifier": {
      "end": 2,
      "greedy": true,
      "kind": 26,
      "start": 2,
      "type": 1,
    },
    "start": 1,
  },
  "end": 3,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > None or many (*) > should correct work with lazy quantifier 1`] = `
{
  "body": {
    "end": 3,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": 6,
      "start": 1,
      "type": "simple",
      "value": "a",
    },
    "kind": 27,
    "quantifier": {
      "end": 3,
      "greedy": false,
      "kind": 26,
      "start": 2,
      "type": 1,
    },
    "start": 1,
  },
  "end": 4,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > None or single (?) > should be parsed 1`] = `
{
  "body": {
    "end": 2,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": 6,
      "start": 1,
      "type": "simple",
      "value": "a",
    },
    "kind": 27,
    "quantifier": {
      "end": 2,
      "greedy": true,
      "kind": 26,
      "start": 2,
      "type": 0,
    },
    "start": 1,
  },
  "end": 3,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > None or single (?) > should correct work with lazy quantifier 1`] = `
{
  "body": {
    "end": 3,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": 6,
      "start": 1,
      "type": "simple",
      "value": "a",
    },
    "kind": 27,
    "quantifier": {
      "end": 3,
      "greedy": false,
      "kind": 26,
      "start": 2,
      "type": 0,
    },
    "start": 1,
  },
  "end": 4,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > Exact > should parse repetition count 1`] = `
{
  "body": {
    "end": 4,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": 6,
      "start": 1,
      "type": "simple",
      "value": "a",
    },
    "kind": 27,
    "quantifier": {
      "end": 4,
      "from": 2,
      "greedy": true,
      "kind": 26,
      "start": 2,
      "to": undefined,
      "type": 3,
    },
    "start": 1,
  },
  "end": 5,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > Exact > should parse unclosed range as chars 1`] = `
{
  "body": {
    "end": 3,
    "expressions": [
      {
        "charCode": 97,
        "end": 1,
        "kind": 6,
        "start": 1,
        "type": "simple",
        "value": "a",
      },
      {
        "charCode": 123,
        "end": 2,
        "kind": 6,
        "start": 2,
        "type": "simple",
        "value": "{",
      },
      {
        "charCode": 50,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": "2",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 4,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > Exact > should parse unclosed range as chars 2`] = `
{
  "body": {
    "end": 3,
    "expressions": [
      {
        "charCode": 97,
        "end": 1,
        "kind": 6,
        "start": 1,
        "type": "simple",
        "value": "a",
      },
      {
        "charCode": 50,
        "end": 2,
        "kind": 6,
        "start": 2,
        "type": "simple",
        "value": "2",
      },
      {
        "charCode": 125,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": "}",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 4,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > From exact to exact > should parse repetition count 1`] = `
{
  "body": {
    "end": 6,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": 6,
      "start": 1,
      "type": "simple",
      "value": "a",
    },
    "kind": 27,
    "quantifier": {
      "end": 6,
      "from": 2,
      "greedy": true,
      "kind": 26,
      "start": 2,
      "to": 4,
      "type": 3,
    },
    "start": 1,
  },
  "end": 7,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > From exact to exact > should parse unclosed range as chars 1`] = `
{
  "body": {
    "end": 5,
    "expressions": [
      {
        "charCode": 97,
        "end": 1,
        "kind": 6,
        "start": 1,
        "type": "simple",
        "value": "a",
      },
      {
        "charCode": 123,
        "end": 2,
        "kind": 6,
        "start": 2,
        "type": "simple",
        "value": "{",
      },
      {
        "charCode": 50,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": "2",
      },
      {
        "charCode": 44,
        "end": 4,
        "kind": 6,
        "start": 4,
        "type": "simple",
        "value": ",",
      },
      {
        "charCode": 52,
        "end": 5,
        "kind": 6,
        "start": 5,
        "type": "simple",
        "value": "4",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 6,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > From exact to exact > should parse unclosed range as chars 2`] = `
{
  "body": {
    "end": 5,
    "expressions": [
      {
        "charCode": 97,
        "end": 1,
        "kind": 6,
        "start": 1,
        "type": "simple",
        "value": "a",
      },
      {
        "charCode": 50,
        "end": 2,
        "kind": 6,
        "start": 2,
        "type": "simple",
        "value": "2",
      },
      {
        "charCode": 44,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": ",",
      },
      {
        "charCode": 52,
        "end": 4,
        "kind": 6,
        "start": 4,
        "type": "simple",
        "value": "4",
      },
      {
        "charCode": 125,
        "end": 5,
        "kind": 6,
        "start": 5,
        "type": "simple",
        "value": "}",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 6,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > From exact to unlimited > should parse repetition count 1`] = `
{
  "body": {
    "end": 5,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": 6,
      "start": 1,
      "type": "simple",
      "value": "a",
    },
    "kind": 27,
    "quantifier": {
      "end": 5,
      "from": 2,
      "greedy": true,
      "kind": 26,
      "start": 2,
      "to": Infinity,
      "type": 3,
    },
    "start": 1,
  },
  "end": 6,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > From exact to unlimited > should parse unclosed range as chars 1`] = `
{
  "body": {
    "end": 4,
    "expressions": [
      {
        "charCode": 97,
        "end": 1,
        "kind": 6,
        "start": 1,
        "type": "simple",
        "value": "a",
      },
      {
        "charCode": 123,
        "end": 2,
        "kind": 6,
        "start": 2,
        "type": "simple",
        "value": "{",
      },
      {
        "charCode": 50,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": "2",
      },
      {
        "charCode": 44,
        "end": 4,
        "kind": 6,
        "start": 4,
        "type": "simple",
        "value": ",",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 5,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > From exact to unlimited > should parse unclosed range as chars 2`] = `
{
  "body": {
    "end": 4,
    "expressions": [
      {
        "charCode": 97,
        "end": 1,
        "kind": 6,
        "start": 1,
        "type": "simple",
        "value": "a",
      },
      {
        "charCode": 50,
        "end": 2,
        "kind": 6,
        "start": 2,
        "type": "simple",
        "value": "2",
      },
      {
        "charCode": 44,
        "end": 3,
        "kind": 6,
        "start": 3,
        "type": "simple",
        "value": ",",
      },
      {
        "charCode": 125,
        "end": 4,
        "kind": 6,
        "start": 4,
        "type": "simple",
        "value": "}",
      },
    ],
    "kind": 5,
    "start": 1,
  },
  "end": 5,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;

exports[`Repetition > Range > should correct work with lazy quantifier 1`] = `
{
  "body": {
    "end": 5,
    "expression": {
      "charCode": 97,
      "end": 1,
      "kind": 6,
      "start": 1,
      "type": "simple",
      "value": "a",
    },
    "kind": 27,
    "quantifier": {
      "end": 5,
      "from": 2,
      "greedy": false,
      "kind": 26,
      "start": 2,
      "to": undefined,
      "type": 3,
    },
    "start": 1,
  },
  "end": 6,
  "flags": "",
  "kind": 0,
  "start": 0,
}
`;
